docs/design.md

```diff
@@ -0,0 +1,872 @@
+
+Sample `BackupInstance` crd:
+
+```yaml
+apiVersion: stash.appscode.com/v1alpha2
+kind: BackupInstance
+metadata:
+  name: backup-volume-demo-instance
+  namespace: demo
+spec:
+  # targetBackupConfiguration indicates the BackupConfiguration crd of respective target that we want to backup
+  targetBackupConfiguration:
+    name: backup-volume-demo
+status:
+  observedGeneration: 239844#2
+  phase: Succeed
+  stats:
+    snapshot: 40dc1520
+    size: 1.720 GiB
+    uploaded: 1.200 GiB # upload size can be smaller than original file size if there are some duplicate files
+    fileStats:
+      new: 5307
+      changed: 0
+      unmodified: 0
+```
+
+
+
+## Recover Workload Data
+
+User will be able to recover backed up data  either into a separate volume or into the same workload from where the backup was taken. Here, is an example for recovering into same workload.
+
+**What user have to do?**
+
+- Create a `Recovery` crd pointing `recoverTo` field to the workload.
+
+Sample `Recovery` crd to recover into same workload:
+
+```yaml
+apiVersion: stash.appscode.com/v1alpha2
+kind: Recovery
+metadata:
+  name: recovery-database-demo
+  namespace: demo
+spec:
+  repository:
+    name: stash-backup-repo
+    namespace: demo
+  paths:
+  - /source/data
+  recoverTo:
+    # indicates the running workload where we want to recover
+    workload:
+      apiVersion: apps/v1
+      kind: Deployment
+      name: stash-demo
+```
+
+**How it will work?**
+
+- When Stash will find a `Recovery` crd created to recover into a workload, it will inject a `init-container` to the targeted workload.
+- Then, it will restart the workload.
+- The `init-container` will recover data inside the workload.
+
+> **Warning:** Recover in same workload require to restart the workload. So, there will be downtime of the workload.
+
+## Backup Volume
+
+User will be also able to backup stand-alone volumes. This is useful for `ReadOnlyMany` or `ReadWriteMany` type volumes.
+
+**What user have to do?**
+
+- Create a `Repository` crd for respective backend.
+
+- Create a `BackupConfiguration` crd pointing `target` field to the volume.
+
+Sample `BackupConfiguration` crd to backup a PVC:
+
+```yaml
+apiVersion: stash.appscode.com/v1alpha2
+kind: BackupConfiguration
+metadata:
+  name: backup-volume-demo
+  namespace: demo
+spec:
+  schedule: '@every 1h'
+  # stashTemplate indicates StashTemplate crd to use for backup the target volume.
+  # stash will create some default StashTemplate  while install to backup/recover various resources.
+  # user can also crate their own StashTemplate to customize backup/recovery
+  stashTemplate: volumeBackup
+  # repository refers to the Repository crd that hold backend information
+  repository:
+    name: stash-backup-repo
+  # target indicate the target workload that we want to backup
+  target:
+    volume:
```

Comment:
```
spec:
  target:
    ref:
      apiVersion:
      kind: PersistentVolumeClaim
      name: <---->
   mountPath: </xyz> # required
   paths: # not supported
```
- We don't allow user to directly backup a PV. They must create a matching PVC in the same namespace as `BackupConfiguration`. This ensures that users can't bypass RBAC in the cluster via Stash.